### ISSUE TITLE
`SpinLock::irq_disabled` may handle irq disable state incorrectly

### ISSUE THREAD

**Issue by StevenJiang1110** [asterinas/asterinas#1120]
Similar problem as https://github.com/Rust-for-Linux/linux/issues/998.

# Reproduce

The simplest case to reproduce the bug is

```rust
let lock1 = SpinLock::new(());
let lock2 = SpinLock::new(());

// Assume interrupt is enabled here
let guard1 = lock1.lock_irq_disabled();
let guard2 = lock2.lock_irq_disabled();

// Interrupt is diabled 

drop(guard1);

// Interrupt is enabled here even guard2 is not dropped!!!!
// If some intrrupt comes here, there will be dead lock problem.

drop(guard2);
```

# Solution
Any user must release lock in the reverse order of acquiring locks, otherwise deadlock may be introduced. 

Since most code relies on compiler to insert `drop` to deal with lock releasing, and compiler will always `drop` newly created object first, it won't be a big problem for current codebase. 

But any code that tries to manually release lock should still obey the above rule and be audited.

**Comment by tatetian** [comment]
> Any user must release lock in the reverse order of acquiring locks, otherwise deadlock may be introduced.

I don't think this claim is correct. Generally speaking, the locking order must be consistent. But the unlocking order is unimportant because unlocking is unconditional.
